plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java

```diff
@@ -179,82 +253,224 @@ public boolean isEmptyConnection() {
+			final FBNetworkElement sourceFB = FordiacMarkerHelper.createErrorMarkerFB(getSourceFbName());
+			setSrcInterfaceList(sourceFB.getInterface());
+			importer.getFbNetwork().getNetworkElements().add(sourceFB);
+			sourceFB.setName(NameRepository.createUniqueName(sourceFB, sourceFB.getName()));
+			createErrorMarkerInterface(false);
 		}
 
-		public void setDestination(final String destination) {
-			this.destination = destination;
+		private void handleMissingConnectionDestination() {
+			// check if there is already one
+			final FBNetworkElement destinationFb = FordiacMarkerHelper.createErrorMarkerFB(getDestFbName());
+			setDestInterfaceList(destinationFb.getInterface());
+			importer.getFbNetwork().getNetworkElements().add(destinationFb);
+			destinationFb.setName(NameRepository.createUniqueName(destinationFb, destinationFb.getName()));
+			createErrorMarkerInterface(true);
 		}
 
-		public void setDestinationEndpoint(final IInterfaceElement destinationEndpoint) {
-			this.destinationEndpoint = destinationEndpoint;
+		private void handleMissingSrcAndDestEnpoint() {
+			final DataType pinType = determineConnectionType(connection);
+
+			final ErrorMarkerInterface srcEndpoint = FordiacErrorMarkerInterfaceHelper
+					.createErrorMarkerInterface(pinType, getSourcePinName(), false, getSrcInterfaceList());
+			final ErrorMarkerInterface destEndpoint = FordiacErrorMarkerInterfaceHelper
+					.createErrorMarkerInterface(pinType, getDestinationPinName(), true, getDestInterfaceList());
+			connection.setSource(srcEndpoint);
+			connection.setDestination(destEndpoint);
 		}
 
-		public void setSource(final String source) {
-			this.source = source;
+		private DataType determineConnectionType(final Connection connection) {
+			if (connection instanceof EventConnection) {
+				return EventTypeLibrary.getInstance().getType(null);
+			}
+			if (connection instanceof AdapterConnection) {
+				final AdapterTypeEntry entry = importer.getTypeLibrary().getAdapterTypeEntry("ANY_ADAPTER"); //$NON-NLS-1$
+				if (null != entry) {
+					return entry.getType();
+				}
+				// we don't have an any_adapter return generic adapter, may not be reparable
+				return LibraryElementFactory.eINSTANCE.createAdapterType();
+			}
+			if (connection instanceof DataConnection) {
+				return IecTypes.GenericTypes.ANY;
+			}
+			return null;
 		}
 
-		public void setSourceEndpoint(final IInterfaceElement sourceEndpoint) {
-			this.sourceEndpoint = sourceEndpoint;
+		private ErrorMarkerInterface createErrorMarkerInterface(final boolean isInput) {
+			final IInterfaceElement oppositeEndpoint = isInput ? getSourceEndpoint() : getDestinationEndpoint();
+
+			// we need a special treatment for FB's that lost their type
+			if (oppositeEndpoint == null) {
+				getConnectionState().add(ConnectionState.MISSING_TYPE);
+				return null;
+			}
+
+			DataType type = oppositeEndpoint.getType();
+			if (type == null) {
+				type = IecTypes.GenericTypes.ANY;
```

Comment:
Should this not also use `determineConnectionType(...)` above to properly handle event or adapter connections?